File: tests/ut/cpp/python_input/gtest_input/optimizer/renormalize_test.py
# Copyright 2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
""" renormalize_test """
from mindspore import Tensor
from mindspore import dtype as mstype
from mindspore.ops import functional as F
from mindspore.ops import operations as P
from mindspore import ops
import mindspore as ms


class FnDict:
    def __init__(self):
        self.fn_dict = {}

    def __call__(self, fn):
        self.fn_dict[fn.__name__] = fn

    def __getitem__(self, name):
        return self.fn_dict.get(name)


# Test ut for function: Renormalize.
def test_renormalize(tag):
    """
    Feature: Renormalize.
    Description: test cases for renormalize.
    Expectation: graph can be renormalized and no exception raised.
    """
    fns = FnDict()
    x = Tensor([1], mstype.int32)
    y = Tensor([1, 2], mstype.int32)
    pow_ops = P.Pow()

    def test_poly_delay_specialize_ut():
        def function_h():
            pow_res = pow_ops(x, x)

            def function_g(param_x):
                return pow(pow_res, param_x)

            return F.make_tuple(pow_res, function_g)

        def function_f():
            h_out = function_h()
            h_forward_out = F.tuple_getitem(h_out, 0)
            g = F.tuple_getitem(h_out, 1)

            def function_k():
                kout1 = g(x)
                kout2 = g(y)
                kout = F.depend(kout1, kout2)
                return kout

            return F.make_tuple(h_forward_out, function_k)

        out = function_f()
        forward_out = F.tuple_getitem(out, 0)
        closure_out = F.tuple_getitem(out, 1)
        closure_out_tensor = closure_out()
        return F.add(forward_out, closure_out_tensor)
    ######################################################

    def test_ignore_flag_with_twice_call_if():
        ll = (Tensor([1]), Tensor([2]), Tensor([3]))

        def func(index, m, n):
            func_out = Tensor([0])
            if ops.less(m, n):
                func_out = ms.ops.tuple_getitem(ll, index)
            return func_out
        x1 = pow_ops(x, x)
        y1 = pow_ops(x, x)
        out1 = func(0, x1, y1)
        out2 = func(1, x1, y1)
        return out1, out2
    ######################################################
    # Add test_poly_delay_specialize_ut to fn dict.
    fns(test_poly_delay_specialize_ut)
    fns(test_ignore_flag_with_twice_call_if)

    return fns[tag]

File: tests/ut/cpp/optimizer/renormalize_test.cc
/**
 * Copyright 2022 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <memory>

#include "common/common_test.h"
#include "mindspore/core/ops/sequence_ops.h"
#include "common/py_func_graph_fetcher.h"

#include "ir/anf.h"
#include "ir/visitor.h"
#include "ir/func_graph_cloner.h"
#include "frontend/optimizer/optimizer.h"
#include "frontend/optimizer/opt.h"
#include "frontend/optimizer/anf_visitor.h"
#include "frontend/optimizer/irpass.h"
#include "frontend/optimizer/irpass/arithmetic_simplify.h"
#include "pipeline/jit/ps/action.h"

#include "include/common/debug/draw.h"
#include "frontend/operator/ops.h"
#include "include/common/utils/cse.h"
#include "include/common/utils/convert_utils.h"

namespace mindspore {
namespace opt {
class TestRenormalize : public UT::Common {
 public:
  TestRenormalize() : getPyFun("gtest_input.optimizer.renormalize_test", true) {}
  void SetUp() {}

 public:
  UT::PyFuncGraphFetcher getPyFun;
};

// Feature: Specialize.
// Description: If a poly node's parent are not specialized, poly node should be delay specialized.
// Expectation: graph can be executed and no exception raised.
TEST_F(TestRenormalize, TestPolyDelaySpecialize) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_renormalize", "test_poly_delay_specialize_ut");
  ASSERT_TRUE(nullptr != test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  std::vector<AbstractBasePtr> args_spec;
  pipeline::Renormalize(res, test_graph, args_spec);
  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}

// Feature: Static analysis of control flow.
// Description: IgnoreValue flag should not be tagged when a function called twice if the function is header of 'if'.
// Expectation: No tuple-getitem exist in specialized graph.
TEST_F(TestRenormalize, TestIgnoreValueTag) {
  FuncGraphPtr test_graph = getPyFun.CallAndParseRet("test_renormalize", "test_ignore_flag_with_twice_call_if");
  ASSERT_TRUE(nullptr != test_graph);
  pipeline::ResourcePtr res = std::make_shared<pipeline::Resource>();
  std::vector<AbstractBasePtr> args_spec;
  auto specialized_fg = pipeline::Renormalize(res, test_graph, args_spec);
  const auto all_nodes = TopoSort(specialized_fg->get_return(), SuccDeeperSimple, AlwaysInclude);
  auto exist_tuple_getitem = std::any_of(all_nodes.cbegin(), all_nodes.cend(), [](const AnfNodePtr &node) {
    return IsPrimitiveCNode(node, prim::kPrimTupleGetItem);
  });
  if (exist_tuple_getitem) {
    DumpIR("test_ignore_flag_with_twice_call_if_error_graph.ir", specialized_fg);
    MS_LOG(ERROR) << "Specialize graph failed, please see the wrong graph in "
                     "'test_ignore_flag_with_twice_call_if_error_graph_0000.ir'";
  }
  ASSERT_EQ(exist_tuple_getitem, false);
  abstract::AnalysisResultCacheMgr::GetInstance().Clear();
  abstract::AnalysisContext::ClearContext();
}
}  // namespace opt
}  // namespace mindspore
